Fix broken watches

tnozicka opened the pull request:
The main reason for having this is that `watch.Until` can exit before the timeout is reached. There are several issues about it in Kubernetes repo already (linked below). It is caused by the watcher being closed prematurely. The usual causes are API timeout or ETCD timeout for watch. This is destined to cause a lot of test flakes! And `watch.Until` already had to be removed from some tests. Also I think this was an issue with `kubectl rollout status --timeout=...` so not just tests.

The idea behind this PR is to inspect all the events being processed and keeping track of last resourceVersion of its objects. If the watch fails it will restart the watch from the last resourceVersion.
This is all wrapped in object RetryWatcher implementing watch.Interface so it looks like a regular watcher and the reader won't even notice those restarts on background.

This PR is not fixing `watch.Until` itself but only providing you with a watcher wrapper to make it behave properly. The reason why I haven't plumbed it into `watch.Until` itself is that it would require changing it's arguments which is a fairly big change I guess - but still, it's broken now. It would need to accept WatchFunc instead of a watcher. Otherwise the change would be just in using RetryWatcher internally in `watch.until` and fixing the code that calls it. Please let me know your opinion here about fixing that as well! Doing this in `watch.Until` would be probably best; it would fix all the related issues... like https://github.com/kubernetes/kubernetes/issues/40224

Please note that this fixes most of the problems but not all. There can be a situation when (between the background restarts) the lastResourceVersion is not in ETCD cache anymore causing it to fail. I guess the only other path forward would be to use Informers but that's kind of heavy and would make the user experience worse.

### Example Usage
```go
initialResourceVersion := ""
watch.Until(10*time.Minute, watch.WithRetry(initialResourceVersion, func(rv string) watch.Interface {
	watcher, err := c.CoreV1().Secrets(secret.Namespace).Watch(metav1.SingleObject(metav1.ObjectMeta{Name: secret.Name, ResourceVersion: rv}))
	if err != nil {
		t.Fatalf("Failed to create watcher on Secret: %v", err)
	}
	return watcher
}))
```

Related issues and PRs about broken watch.Until:
- https://github.com/kubernetes/kubernetes/pull/41112
- https://github.com/kubernetes/kubernetes/pull/47756
- https://github.com/kubernetes/kubernetes/issues/40224
- https://github.com/kubernetes/kubernetes/pull/33942#issuecomment-305817825
- https://github.com/kubernetes/kubernetes/issues/31345
- https://github.com/kubernetes/kubernetes/issues/39879#issuecomment-278059099
- https://github.com/openshift/origin/pull/14936#issuecomment-311957325

(Feel free to link more.)

ccing few people from previous issues that might be interested:
@smarterclayton @mfojtik @kargakis @soltysh @janetkuo @liggitt @wojtek-t 

**Release note**:
```release-note
Fix premature timeouts/failures in commands that wait using WATCH
```

Split tasks (to go in smaller PRs):
 - [x] Move `"k8s.io/apimachinery/pkg/watch".Until` to `"k8s.io/client-go/tools/watch".UntilWithoutRetry` (https://github.com/kubernetes/kubernetes/pull/66906)
 - [x] Introduce `UntilWithSync` (https://github.com/kubernetes/kubernetes/pull/66971)
 - [ ] Introduce `Until` (with retry)
 - [ ] Individual PRs to switch kubectl and other places to either `UntilWithSync` or `Until`
   - [ ] `kubectl rollout status` - https://github.com/kubernetes/kubernetes/pull/67817

liggitt commented:
Glad to see this being worked on, will take a look.

My preference would be to sweep calls the watch.Until/cache.ListWatchUntil and actually plumb the info needed to handle both the request timeout and the "too old version" errors. Do we know what changes would be needed to address both of those?

Looks like we have aboue 8 uses in non-test code, and about 15 in test code. That seems worth sweeping/fixing.

wojtek-t commented:
+1 to what @liggitt wrote

liggitt commented:
> I guess the only other path forward would be to use Informers but that's kind of heavy and would make the user experience worse.

Given how few call sites there are, it might actually make sense to switch the places using this to use informers, and make it easier to set up an informer for a single object in the process.

kargakis commented:
/ok-to-test

tnozicka commented:
I have been thinking about this more in follow-up to your comments and I have realized there are 2 distinct use cases here and I don't think either informers or restating watcher can solve both.

1. Continuously watching a state of the world
Usual case for controller or commands like `oc rollout status`. This is best covered by an informer to ensure it can recover from watch errors (even too old RV) by re-listing.

2. Continuously watching a stream of events with a guarantee of not missing any and receiving them in the same order as they happened
The main area here would be tests where you need to reason and to validate sequence of actions performed by controllers to validate something happens after another change. I don't think that informers can help here since they do list+watch and will loose events while doing list. On the other hand restarting watch using RV will work for this case.

So I think we need RetryWatcher at least for tests and some cases like `oc rollout status` would be better handled by an informer.

> Looks like we have aboue 8 uses in non-test code, and about 15 in test code. That seems worth sweeping/fixing.

Well there are also many other projects out there using Kubernetes as a Go package. But I would guess that's an easy change for them and good projects vendor anyways. So I will bake it into `watch.Until`

I didn't know about cache.ListWatchUntil but looking at uses of it I guess it would be better switched to an informer with timeout

liggitt commented:
> Continuously watching a stream of events with a guarantee of not missing any and receiving them in the same order as they happened

that's not actually possible to absolutely guarantee. if your watch is dropped, and re-establishing from last seen RV answers with "too old resource version", and a refetch (re-get or re-list) returns a newer RV than the last one you observed, that means you missed events.

I think as many callers as possible should be converted to informers. If there are specific tests that cannot be converted to an informer, let's consider those individually, with the understanding that expecting the behavior you describe leaves us open to flakes.

tnozicka commented:
>that's not actually possible to absolutely guarantee. if your watch is dropped, and re-establishing from last seen RV answers with "too old resource version", and a refetch (re-get or re-list) returns a newer RV than the last one you observed, that means you missed events.

@liggitt I agree with what you are saying here that's why informer won't cover this. On the other hand RetryWatcher will and it shouldn't loose any events; in case of too old resource version it rather fails. As this is use case mainly for tests the too old RV will occur only if something else is wrong (usually DOS or badly configured etcd)

> I think as many callers as possible should be converted to informers. If there are specific tests that cannot be converted to an informer, let's consider those individually, with the understanding that expecting the behavior you describe leaves us open to flakes.

I think in tests we can safely configure etcd to have big enough cache and we run them also sequentially I think so it would be hard to see cases of DOS to cause too old resource version.

tnozicka commented:
@liggitt here is a use real world use case for 2. where it actually revealed races that would have been missed without watching continuous stream of events and likely would have been mistaken for flakes - https://github.com/openshift/origin/pull/14936#issuecomment-311960928

ash2k commented on staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go:
```diff
@@ -0,0 +1,151 @@
+func (rw *RetryWatcher) receive() {
+	for {
+		// We need to wrap this code in a function so the old watchers are freed by defer between retries
+		done := func() bool {
```
Why not make it a method? 1 less allocation, reads better and the loop can be written much nicely:
```go
for !rw.receiveInternal() {
}
```

tnozicka replied:
you are right, I must have lost it somewhere, thx :)

ash2k commented on staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go:
```diff
@@ -0,0 +1,151 @@
+			ch := watcher.ResultChan()
+			defer watcher.Stop()
+
+			select {
```
Shouldn't there be a loop around `select`? Otherwise the watch is recreated after each event - is it how it is supposed to work?

tnozicka replied:
you are right, I must have lost it somewhere, thx :)

smarterclayton commented:
Note that starting with 1.6 and beyond the odds of "resource version too
old" are low unless we wait for 5 minutes without a single update.  I don't
think any of our tests are vulnerable to that.  So I would agree that in
certain circumstances (like tests, where the order of operations matters)
that a retry watcher that watches from the last RV received and errors if
it can't continue is a useful tool, BUT it should be really clear that most
people shouldn't use this.  That may imply we want to put it in its own
package, and clearly document when it is appropriate to use it.

On Wed, Aug 9, 2017 at 5:19 AM, Mikhail Mazurskiy <notifications@github.com>
wrote:

> *@ash2k* commented on this pull request.
> ------------------------------
>
> In staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go
> <https://github.com/kubernetes/kubernetes/pull/50102#discussion_r132131332>
> :
>
> > +		resultChan:          make(chan Event, 1),
> +	}
> +	go rw.receive()
> +	return rw
> +}
> +
> +// receive reads the result from a watcher, restarting it if necessary.
> +func (rw *RetryWatcher) receive() {
> +	for {
> +		// We need to wrap this code in a function so the old watchers are freed by defer between retries
> +		done := func() bool {
> +			watcher := rw.watchFunc(rw.lastResourceVersion)
> +			ch := watcher.ResultChan()
> +			defer watcher.Stop()
> +
> +			select {
>
> Shouldn't there be a loop around select? Otherwise the watch is recreated
> after each event - is it how it is supposed to work?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/50102#pullrequestreview-55177880>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p_ph6jKdueUOjuZIs-vwfuFI_Hz7ks5sWXm0gaJpZM4Os0VC>
> .
>

kargakis commented:
> BUT it should be really clear that most people shouldn't use this. That may imply we want to put it in its own package, and clearly document when it is appropriate to use it.

@smarterclayton the main goal of this PR is to solve the dropped watch issue we have with `kubectl rollout status` - see https://github.com/kubernetes/kubernetes/issues/40224

liggitt commented:
> @smarterclayton the main goal of this PR is to solve the dropped watch issue we have with kubectl rollout status

then it needs to handle an expired watch

tnozicka commented:
@kargakis Actually, after the discussions above, I think the conclusion is that we can't deal with this using one function as the solutions for 1. and 2. are orthogonal.

1. Watching continuous stream of events where the completeness, uniqueness and ordering matters can be achieved only by Watch. Retry watcher makes sure that it can't be broken by API timeout or errors other than ETCD not having it any more; so it can't deal with `RV too old`.

2. Watching the state of the world. List+Watch (an informer) can deal with  `RV too old` but in return it looses some guarantees from 1.

So I will keep this PR open to address 1. (+update it) and we should switch cases like `oc rollout status` to use informers in separate PRs probably preceding this one. I will look into switching `oc rollout status` to start with.

wojtek-t commented:
> So I will keep this PR open to address 1. (+update it) and we should switch cases like oc rollout status to use informers in separate PRs probably preceding this one. I will look into switching oc rollout status to start with.

+1

ncdc commented:
@tnozicka @liggitt @smarterclayton @wojtek-t are there plans to proceed with this?

tnozicka commented:
@ncdc I was overloaded with other bugs lately but I definitely want to move this forward. Actually, I got back to it briefly yesterday when I saw the code freeze date.

My current approach is to switch all the places using watch.Until to either an informer or RetryWatcher in this PR. That should validate the correctness and genericity of the new objects/wrappers/util_functions.

ncdc commented on staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go:
```diff
@@ -0,0 +1,151 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// Watch function is responsible for creating a watcher starting at sinceResourceVersion.
```
WatchFunc is a function that is responsible...

ncdc commented on staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go:
```diff
@@ -0,0 +1,151 @@
+// Watch function is responsible for creating a watcher starting at sinceResourceVersion.
+type WatchFunc func(sinceResourceVersion string) Interface
+
+// Type of an event that will be returned if RetryWatcher fails
```
RetryWatcherError is the type of an event...

tnozicka replied:
done

ncdc commented on staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go:
```diff
@@ -0,0 +1,151 @@
+
+// Interface to get resource version from events.
+// We can't reuse an interface from meta otherwise it would be a cyclic dependency and we need just this one method
+type resourceVersionInterface interface {
```
I'd call this `resourceVersionGetter` but feel free to ignore if you disagree :)

tnozicka replied:
I like it more too :)

ncdc commented on staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go:
```diff
@@ -0,0 +1,151 @@
+	GetResourceVersion() string
+}
+
+// Watchers can be closed e.g. in case API timeout is reached, etcd timeout, ...
```
The Godoc needs to start with the name of the struct (RetryWatcher)

tnozicka replied:
done

ncdc commented on staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go:
```diff
@@ -0,0 +1,151 @@
+	stopChan            chan struct{}
+}
+
+// Creates a new RetryWatcher.
```
NewRetryWatcher...

tnozicka replied:
done

ncdc commented on staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go:
```diff
@@ -0,0 +1,151 @@
+				}
+
+				// We need to inspect the event and get ResourceVersion out of it
+				switch t := event.Type; t {
```
Do you need `t` or can you just switch on `event.Type`?

tnozicka replied:
it was used in one log message, but I'll switch it to use `event.Type` directly there

ncdc commented on staging/src/k8s.io/apimachinery/pkg/watch/retrywatcher.go:
```diff
@@ -0,0 +1,151 @@
+	close(rw.stopChan)
+}
+
+// WithRetry will wrap the watchFunc with RetryWatcher making sure that watcher gets restarted in case of errors.
```
s/will wrap/wraps/

tnozicka replied:
done

sttts commented on staging/src/k8s.io/apimachinery/pkg/watch/tools/retrywatcher.go:
```diff
@@ -0,0 +1,154 @@
+func (obj RetryWatcherError) GetObjectKind() schema.ObjectKind { return schema.EmptyObjectKind }
+
+// DeepCopyObject implements interface.
+func (obj RetryWatcherError) DeepCopyObject() runtime.Object { return obj }
```
Why should this be a runtime.Object?

tnozicka replied:
it is returned as an object in event when error occurs https://github.com/tnozicka/kubernetes/blob/1c33685d5b4c2eaa14a7650dc7f79eeeb195e31c/staging/src/k8s.io/apimachinery/pkg/watch/tools/retrywatcher.go#L95

sttts replied:
Is `RetryWatcherError` exposed to the user of the retry watcher? Wouldn't wrapping `error` make more sense? Or even `Status`.

sttts commented on staging/src/k8s.io/apimachinery/pkg/watch/tools/until.go:
```diff
@@ -0,0 +1,118 @@
+	return lastEvent, nil
+}
+
+// UntilWithRetry can deal with API timeout or lost connections.
```
say first what it does primarily, then which error modes it supports.

tnozicka commented:
/test pull-kubernetes-e2e-gce

tnozicka commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -84,6 +94,71 @@ func NewListWatchFromClient(c Getter, resource string, namespace string, fieldSe
 }
 
+// TODO: really think if it's worth it
+func NewListWatchFromMethods(listerWatcher interface{}, fieldSelector fields.Selector) *ListWatch {
```
I wonder how people feel about this function?

In one way it's really useful and simplifying for its users but hard to maintain.

It also move semantic check from compile time to runtime, but Go just doesn't have templates :/

tnozicka commented:
/retest

tnozicka commented:
/retest
bitbucket.org is having issues
```
W0109 16:33:11.863] abort: HTTP Error 500: Internal Server Error
W0109 16:33:11.863] godep: error downloading dep (bitbucket.org/ww/goautoneg): exit status 255
W0109 16:33:18.479] godep: Error downloading some deps. Aborting restore and check.
W0109 16:33:18.484] !!! Error in ./hack/godep-restore.sh:35
W0109 16:33:18.485]   Error in ./hack/godep-restore.sh:35. 'GOPATH="${GOPATH}:${KUBE_ROOT}/staging" godep restore "$@"' exited with status 1
```

tnozicka commented:
/retest

tnozicka commented:
/retest

tnozicka commented:
Seems like **pull-kubernetes-e2e-gce** has reached its timeout and is killing the test, the rest is green. Any idea on how to raise it?

liggitt commented:
normally the job takes ~40 minutes... need to understand why it is dependably taking more than an hour on this PR

tnozicka commented:
I was hoping we were on the edge, but watches have no doubt the potential to cause it

tnozicka commented:
Hm, the integration test failure is interesting and I think it's a side effect of adding a new integration test. 

To start with I had to raise `nofile` limit because it seems like integration tests are not closing the connection to shared etcd. (etcd was complaining about 80% of fd used)

If I run all the tests locally, it fails on one of them at some point stuck repeating on:
```
unable to sync kubernetes service: Post http://127.0.0.1:33337/api/v1/namespaces: dial tcp 127.0.0.1:33337: connect: connection refused
```
This is kind of appearing on many of the tests as time passes and gets fatal for one at some point.

If I remove the apimachinery integration test I have added, I see this error but integration tests succeed.

If I run from master I see this error but integration tests succeed.

If I run the failed test separately, it succeeds.

tnozicka commented:
/test pull-kubernetes-e2e-kops-aws

tnozicka commented:
/retest

tnozicka commented:
fixed the e2e; only integration remaining - not yet sure why that's failing
/test pull-kubernetes-integration

tnozicka commented:
/test pull-kubernetes-integration

tnozicka commented:
I see integration failed similarly on other PRs (like https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/59786/pull-kubernetes-integration/3240/) so this might not be related to my changes

tnozicka committed:
Switch Kubernetes to use the new functions UntilWithRetry and UntilWithInformer to eliminate watch timeouts

tnozicka commented:
/retest

tnozicka commented:
/test pull-kubernetes-integration

tnozicka commented:
Tests were green; have to rebase again because of new conflicts with master

tnozicka commented:
@liggitt @ncdc @ash2k @wojtek-t @deads2k @soltysh @smarterclayton PTAL

tnozicka commented:
/retest

tnozicka commented:
flake https://github.com/kubernetes/kubernetes/issues/61748
/test pull-kubernetes-integration

tnozicka commented:
/test pull-kubernetes-integration

tnozicka commented:
flake #61748
/test pull-kubernetes-integration

tnozicka commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -93,95 +104,104 @@ func NewFilteredListWatchFromClient(c Getter, resource string, namespace string,
-				continue
+
+			if resErr != nil {
+				glog.Errorf("Failed to List objects: %v", resErr)
```
For some reason I thought informers only log those failures somewhere around loglevel 5 - but by looking the code, it seems like informers log it as glog.Errorf so I can drop this.

tnozicka commented on staging/src/k8s.io/client-go/tools/cache/listwatch_test.go:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	for _, tmptc := range tt {
+		tc := tmptc // we need to copy it for parallel runs
```
yeah, this is copy paste leftover. no need for parallel here.

tnozicka commented:
/test pull-kubernetes-bazel-test

tnozicka commented:
All tests are green again! (Integration flake was fixed in master.)

If you were holding of your review, now is the time :)

tnozicka commented:
/sig api-machinery
/sig cli
/sig apps

ncdc commented on pkg/controller/client_builder.go:
```diff
@@ -25,14 +25,14 @@ import (
 	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
 	restclient "k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/cache"
+	wtools "k8s.io/client-go/tools/watch"
```
watchtools?

ncdc commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -38,6 +38,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/rest"
+	wtools "k8s.io/client-go/tools/watch"
```
watchtools?

ncdc commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -584,7 +586,15 @@ func (options *GetOptions) watch(f cmdutil.Factory, cmd *cobra.Command, args []s
+		// FIXME: Switch to an informer (wtools.UntilWithInformer)
+		// Pure WATCH will fail after some time but definitely on API timeout.
+		// This is not an easy task as the CLI arguments "watch" and "watch-only"
+		// enforce using pure WATCH even in it's description. Those need to be deprecated first.
```
s/it's/its/

ncdc replied:
If we do switch to using an informer, and the watch times out, would you have it do the normal informer behavior or relist (notifying the consumer) and then watching again?

tnozicka replied:
I'd choose a regular informer behaviour. Informer will try to run watch again from the last RV if possible, re-list if not. Continuing watch from last RV in case of a apiserver (or similar) timeout seems preferable. That will give user the best chance to see the exact order of events unless that's no longer possible (too old RV) and notifies him when that happens.

mfojtik commented:
/lgtm

ncdc commented on pkg/kubectl/cmd/rollout/rollout_status.go:
```diff
@@ -20,15 +20,18 @@ import (
+	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	wtools "k8s.io/client-go/tools/watch"
```
watchtools?

ncdc commented:
/hold

I'm reviewing and have some comments

ncdc commented on pkg/kubectl/cmd/rollout/rollout_status.go:
```diff
@@ -110,7 +114,8 @@ func RunStatus(f cmdutil.Factory, cmd *cobra.Command, out io.Writer, args []stri
 	if err != nil {
 		return err
 	}
-	rv, err := mapping.MetadataAccessor.ResourceVersion(obj)
```
Are we losing watching from rv?

ncdc replied:
Without digging into the resource builder code, I would hope/expect this to be the RV of either the list (if doing a list) or the single object (if doing a get). I assume the reason for capturing the RV was to be able to start a watch from the same RV instead of getting all the revisions currently available (anything that hasn't yet been collapsed into an etcd snapshot).

@liggitt any thoughts here?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/watch/filter.go:
```diff
@@ -87,10 +87,14 @@ var _ Interface = &Recorder{}
 	return r
 }
 
+func (r *Recorder) SetInterface(w Interface) {
```
this is a little difficult to reason about... what are the thread safety guarantees around this? if Stop() or ResultChan() was invoked on the previous interface, do events from that interface continue to get recorded in the background?

tnozicka replied:
No thread safety guarantees, I should have made that clear with a comment. This is used in the case where the watcher fails and the RetryWatcher creates a new one to update watcher for the Recorder and because for the first watcher is actually created after the Recorder.

ncdc commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -93,95 +102,96 @@ func NewFilteredListWatchFromClient(c Getter, resource string, namespace string,
-	watchInterface, err := lw.Watch(metav1.ListOptions{ResourceVersion: currResourceVersion})
-	if err != nil {
-		return nil, err
+// List a set of apiserver resources
```
`// List lists ...`

ncdc commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -93,95 +102,96 @@ func NewFilteredListWatchFromClient(c Getter, resource string, namespace string,
-		err = wait.ErrWaitTimeout
-	}
-	return evt, err
+// Watch a set of apiserver resources
```
`// Watch watches ...`

ncdc commented on staging/src/k8s.io/client-go/tools/cache/listwatch_test.go:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	for _, tc := range tt {
+		t.Run("", func(t *testing.T) {
```
If you're using subtests, please give each one a name (in your `tt` struct)

ncdc commented on staging/src/k8s.io/client-go/tools/watch/informerwatcher.go:
```diff
@@ -0,0 +1,60 @@
+	"k8s.io/client-go/tools/cache"
+)
+
+// NewInformerWatcher will IndexedInformer and wrap it into watch Interface
```
will **create an** IndexedInformer?

watch.Interface (period in between)?

ncdc commented on staging/src/k8s.io/client-go/tools/watch/informerwatcher_test.go:
```diff
@@ -0,0 +1,186 @@
+	}
+
+	for _, tc := range tt {
+		t.Run("", func(t *testing.T) {
```
Please use named subtests

ncdc commented on staging/src/k8s.io/client-go/tools/watch/informerwatcher_test.go:
```diff
@@ -0,0 +1,186 @@
+					}
+
+					result = append(result, event)
+				case <-time.After(time.Second * 1):
```
Is this expecting that all watch events will be received in less than 1 second? If so, I'd wonder if this could potentially flake in the Kube testing infra, where lots of tests run in parallel?

ncdc commented on staging/src/k8s.io/client-go/tools/watch/informerwatcher_test.go:
```diff
@@ -0,0 +1,186 @@
+			w.Stop()
+
+			// Wait a bit to see if the informer won't panic
+			time.Sleep(1 * time.Second)
```
Same question here about possible flakiness

wojtek-t replied:
I'm not a fan of this sleep, but I see the motivation.
Ideally, we should wait for underlying watch to be close or sth like that instead of sleeping.

So maybe add a TODO to replace this sleep with something more reliable.

ncdc commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher_test.go:
```diff
@@ -0,0 +1,395 @@
+	}
+
+	for _, tc := range tt {
+		t.Run("", func(t *testing.T) {
```
Please use named subtests

ncdc commented on staging/src/k8s.io/client-go/tools/watch/until.go:
```diff
@@ -0,0 +1,115 @@
+//
+// Warning: Unless you have a very specific use case (probably a special Watcher) don't use this function!!!
+// Warning: This will fail e.g. on API timeouts and/or 'too old resource version' error.
+// Warning: You are most probably looking for a function *UntilWithRetry* or *UntilWithInformer* bellow,
```
typo: below

ncdc reviewed (changes requested):
I made it through a good portion of the PR. I don't have time right now to thoroughly examine all the changes to the e2e & integration tests.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -93,95 +102,96 @@ func NewFilteredListWatchFromClient(c Getter, resource string, namespace string,
-func (lw *ListWatch) Watch(options metav1.ListOptions) (watch.Interface, error) {
-	return lw.WatchFunc(options)
-}
+// NewListWatchFromMethods creates a ListWatch object from a client supporting List and Watch methods.
```
is this change necessary? this doesn't seem like an improvement to me

tnozicka replied:
I've been on both sides of this fence, swearing at Golang, ended up embracing this path. The function is covered by tests so presumably the only time this fails and fails consistently (with panic) is when you give it object not fulfilling the interface. I understand your point, but I wouldn't overestimate the compiler safety. Aren't informers (and others) erasing the type to `interface{}` as well to go around the fact that Golang doesn't have templates. We get other benefits and this is a way of living with the downsides.

tnozicka replied:
If informers precedent wouldn't be enough, we are also loosing compiler safety (in different way) in almost identical function
https://github.com/kubernetes/kubernetes/blob/1ee2ac07c10dd8096d37ace3627bdffe45c2b867/staging/src/k8s.io/client-go/tools/cache/listwatch.go#L74-L94
by treating `resource` as string instead of using the typed client.

wojtek-t replied:
I'm actually with @liggitt on this.
And I think there is difference between the example from above and what you're doing.

In the example above, we may potentially send a request to apiserver for non-existing resource, which will thrn result some error (which we need to handle anyway).

In the code below you're doing a lot of reflect.* operations which will be called at runtime (including panics).
I would really like to avoid getting that path. I would be very afraid of running such code as part of production.

tnozicka replied:
I'll drop it to move this PR forward

wojtek-t replied:
Great - please ping me when this will be ready for next review pass.

liggitt commented on staging/src/k8s.io/client-go/tools/watch/until.go:
```diff
@@ -0,0 +1,115 @@
+// Due to this guarantee there is no way it can deal with 'Resource version too old error'. It will fail in this case.
+// The most frequent usage would be a test where you want to verify exact order of events.
+func UntilWithRetry(timeout time.Duration, initialResourceVersion string, watcherFunc WatcherFunc, conditions ...ConditionFunc) (*watch.Event, error) {
+	return UntilWithoutRetry(timeout, NewRetryWatcher(initialResourceVersion, watcherFunc), conditions...)
```
UntilWithRetry delegating to UntilWithoutRetry reads very strangely...

tnozicka replied:
I am open to better names, this was a result of previous conversations.

liggitt commented on test/integration/apimachinery/watch_restart_test.go:
```diff
@@ -0,0 +1,264 @@
+		for {
+			select {
+			// TODO: get this lower once we figure out how to extend ETCD cache
+			case <-time.After(1000 * time.Millisecond):
```
is this working around the watch cache 1 second restart interval?

liggitt commented on staging/src/k8s.io/client-go/tools/watch/until.go:
```diff
@@ -0,0 +1,115 @@
+// It guarantees you to see all events and in the order they happened.
+// Due to this guarantee there is no way it can deal with 'Resource version too old error'. It will fail in this case.
+// The most frequent usage would be a test where you want to verify exact order of events.
+func UntilWithRetry(timeout time.Duration, initialResourceVersion string, watcherFunc WatcherFunc, conditions ...ConditionFunc) (*watch.Event, error) {
```
do we want to put this in a different testing-oriented package to try to clarify this is not what you should use for non-test cases?

liggitt commented on staging/src/k8s.io/client-go/tools/watch/informerwatcher.go:
```diff
@@ -0,0 +1,60 @@
+
+// NewInformerWatcher will IndexedInformer and wrap it into watch Interface
+// so you can use it anywhere normal watcher from Watch method could be used.
+func NewInformerWatcher(lw cache.ListerWatcher, objType runtime.Object, resyncPeriod time.Duration) watch.Interface {
```
do we want to coerce an informer into a watcher interface? it makes it easy to think you are supposed to be observing every watch event when you aren't really... is that the promise of the watch interface?

tnozicka commented:
I had to rebase to solve hidden conflict on `masterConfig.ExtraConfig.EnableCoreControllers = false` which has been removed from the extra config in https://github.com/kubernetes/kubernetes/pull/62030

tnozicka commented:
flake: https://github.com/kubernetes/kubernetes/issues/62356
/retest

tnozicka commented:
comments addressed, feel free to take another look

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -93,95 +102,96 @@ func NewFilteredListWatchFromClient(c Getter, resource string, namespace string,
+// NewListWatchFromMethods creates a ListWatch object from a client supporting List and Watch methods.
+// Because Golang doesn't have templates/generics to allow us to accept an object implementing List+Watch interface but differing
+// in return types the client's type is erased to in interface{} and checked in runtime using reflection.
+func NewListWatchFromMethods(listerWatcher interface{}, userListOptions metav1.ListOptions) *ListWatch {
```
I would really like to see listerWatcher being of type ListerWatcher. What is the exact problem that you have with that?

tnozicka replied:
I think the issue was that ListerWatcher return types don't match as List and Watch method return types for every object differ there from runtime.Object, so you can't put the method there directly. Anyways I am dropping it.

wojtek-t commented on staging/src/k8s.io/client-go/tools/watch/until.go:
```diff
@@ -0,0 +1,115 @@
+				continue
+			}
+		}
+	ConditionSucceeded:
```
Can we avoid those named statements?
It should be easy to avoid that by putting that into function here.

tnozicka replied:
I think this is just the old `watch.Until` moved, not a new code.

wojtek-t commented on staging/src/k8s.io/client-go/tools/watch/until.go:
```diff
@@ -0,0 +1,115 @@
+// Conditions are satisfied sequentially so as to provide a useful primitive for higher level composition.
+// A zero timeout means to wait forever.
+//
+// Warning: Unless you have a very specific use case (probably a special Watcher) don't use this function!!!
```
Do you see any usecase for it?

Why do we even make it public then?

ncdc replied:
I agree with @wojtek-t. If we expose a function, someone will eventually use it.

tnozicka replied:
I we keep it we don't have to switch all the places at once. There are also many cases not covered by this PR, temporarily switched to pooling, that might raise the need. (Like creating the watcher in advance at some specific point to capture RV, although you can do it with List - unless you have RBAC only for Watch.) I don't feel strong about it though.

wojtek-t replied:
If that's about making this PR smaller and solving that in a separate one, I'm fine with that.
But then instead of making warning, please add an explicit TODO to make it private function after migrating all its users to UntilWithRetry

tnozicka replied:
We can't use it in integration tests if that's private...

wojtek-t replied:
And we shouldn't use that in integration tests.
But as I wrote, I'm fine with TODO to get rid of usages of that later..

tnozicka replied:
> And we shouldn't use that in integration tests.

@wojtek-t could you elaborate on that? This PR actually add one valid use case to integration tests `test/integration/apimachinery/watch_restart_test.go`

tnozicka replied:
Added a TODO below

wojtek-t commented:
@liggitt @ncdc @tnozicka - merging that PR is blocking fixing the scalability-related flakes -
 https://github.com/kubernetes/kubernetes/issues/62064
Can we somehow help with it to speed it up?

I think the crucial part is to agree on changes to:
`staging/src/k8s.io/client-go/tools/watch/until.go` file
I added some comments there few days ago, so it would be good to agree/disagree on those.

tnozicka commented:
ok, hope it last a while, had to rebase a second time in an hour because of new conflicts.

@liggitt @wojtek-t reflect magic removed

tnozicka commented:
/test pull-kubernetes-integration

wojtek-t reviewed (commented):
A couple comments - but overall looks very good now.
Thanks a lot for changing that!

wojtek-t commented on staging/src/k8s.io/client-go/tools/watch/until.go:
```diff
@@ -0,0 +1,115 @@
+	return lastEvent, nil
+}
+
+// UntilWithRetry wraps the watchFunc with RetryWatcher making sure that watcher gets restarted in case of errors.
```
@ncdc @liggitt - thought? [I personally vote for just `Until`]

ncdc replied:
+1 on `Until`

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/watch/watch.go:
```diff
@@ -268,3 +268,49 @@ func (f *RaceFreeFakeWatcher) Action(action EventType, obj runtime.Object) {
+	defer pw.mutex.Unlock()
+	if !pw.stopped {
+		pw.stopped = true
+		close(pw.stopCh)
```
OK - fair enough.

wojtek-t commented on staging/src/k8s.io/client-go/tools/watch/informerwatcher.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+// NewInformerWatcher will create an IndexedInformer and wrap it into watch.Interface
+// so you can use it anywhere normal watcher from Watch method could be used.
```
Yeah - that sounds fine to me.

wojtek-t commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,178 @@
+					Type:   watch.Error,
+					Object: RetryWatcherError{fmt.Sprintf("__internal__: RetryWatcher failed to recognize Event type %q", event.Type)},
+				}
+				// We have to abort here because this might cause lastResourceVersion inconsistency!
```
I agree we shouldn't continue.
I was just asking about the phrasing of this comment. Maybe we should clarify it a bit more.

wojtek-t commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,178 @@
+			// We need to inspect the event and get ResourceVersion out of it
+			switch event.Type {
+			case watch.Added, watch.Modified, watch.Deleted:
+				metaObject, ok := event.Object.(resourceVersionGetter)
```
In general yes.
But, can we ensure that those won't be coming (via some convoluted path) from informer?
It may not be possible now, but we may add some strange path in the future...

tnozicka replied:
I started implementing this as it seemed pretty easy, but then I've found out that the interface of event requires `runtime.Object` and `cache.DeletedFinalStateUnknown` doesn't implement this interface. At that point future proofing get kind of weird with the extra `interface{}` cast.
```golang
tombstone, ok := event.Object.(interface{}).(cache.DeletedFinalStateUnknown)
```

I don't think that's a good fit.

liggitt replied:
agree we don't need to check that here when dealing with something that conforms to the watcher interface, but we do in NewInformerWatcher

tnozicka replied:
I've added the check there, thx

serngawy commented:
could be related

https://github.com/kubernetes/kubernetes/issues/62549

wojtek-t commented:
@tnozicka - friendly ping for this PR - I would really like to see this in for 1.11

tnozicka commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,178 @@
+			glog.V(5).Info("Stopping RetryWatcher.")
+			return
+		default:
+			done := rw.doReceive()
```
you likely meant
```
if done := rw.doReceive(); done {
  return
}
```

For better readability I generally prefer to keep this on 2 lines so every `if` statement is directly followed by the condition, not initialization. With longer statements you sometimes even have to scroll to see the condition using the 1 line `if`.

tnozicka committed:
Add apimachinery tools to deal with broken watches:

 - watch.RetryWatcher
 - retry.RetryWatcher
 - informer.NewInformerWatcher
 - watch.UntilWithRetry
 - watch.UntilWithInformer
 - moves watch.Until to watch.UntilWithoutRetry with a big warning not to use it

 Remove cache.ListWatchUntil because there is no use for it anymore

tnozicka commented:
/retest

tnozicka commented:
rebased for like a 5th time in 24h, updated to address comments, enhanced error handling for RetryWatcher's watchFunc

can't update Godeps for now as that script is broken - https://github.com/kubernetes/kubernetes/issues/63166

liggitt commented on staging/src/k8s.io/client-go/tools/watch/informerwatcher.go:
```diff
@@ -0,0 +1,60 @@
+		DeleteFunc: func(obj interface{}) {
+			ch <- watch.Event{
+				Type:   watch.Deleted,
+				Object: obj.(runtime.Object),
```
DeleteFunc can be passed a `cache.DeletedFinalStateUnknown` object as well as a runtime.Object

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,179 @@
+						Type:   watch.Error,
+						Object: RetryWatcherError{"__internal__: RetryWatcher: doesn't support resourceVersion"},
+					}
+					// We have to abort here because this might cause lastResourceVersion inconsistency by skipping a potential RV with valid data!
```
we don't actually have to abort, we just can't restart the watch if it drops

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		rw.resultChan <- watch.Event{
+			Type:   watch.Error,
+			Object: RetryWatcherError{fmt.Sprintf("RetryWatcher: watcherFunc failed: %v", err)},
```
using `apierrors.NewInternalError(fmt.Errorf("RetryWatcher: watcherFunc failed: %v", err)).Status()` instead of RetryWatcherError would be more natural and more likely to be handled/surfaced by watchers (metav1.Status errors are typically what is returned with watch.Error events)

wojtek-t commented:
Can we resurrect this PR? I would really like to have this happened.

tnozicka commented:
Yeah, I am gonna give it another try. 

(I've been taking most of my PTO for this year for personal reasons, but I am back. Sorry about the delays.)

tnozicka commented:
While rebasing it on the latest master and rewriting the new `kubectl wait` I have came to conclusion that with with the `UntilWithInformer` we need a concept of precondition. A particular use case would be waiting for deletion - if the object would be deleted before the informer does List, so it's not in its cache yet, you wouldn't see the delete event. With a precondition `NewInformerWatcher` would wait for caches to sync, checked the pre-condition (e.g. does the object exist in cache by doing a List on the informer) and either finished or continued the regular path.

wojtek-t commented:
friendly ping on this one

re your idea above - that sounds reasonable, though I don't think it's prerequiste for this PR itself.

tnozicka commented:
/test pull-kubernetes-integration

tnozicka committed:
Enhance apimachinery/pkg/api/errors.APIStatus.Status() interface to return a pointer and fulfill runtime.Object interface

tnozicka commented:
Comments addressed, new concepts are in, tests were green (before commit clean up). (Code might need some tiny polishing still.)

Integration test seems flaky but I need to look into that yet.

tnozicka commented:
/retest

tnozicka commented:
/retest

tnozicka commented:
pull-kubernetes-e2e-kops-aws - not sure about that job, sometimes [nodes fail to start](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/50102/pull-kubernetes-e2e-kops-aws/98602/) because it can't connect to apiserver, but there are no apiserver logs to check that; sometimes it fails with RV too old on StatefulSet test which is actually a manifestation of [waiting 5m and ~30s for the pod to get up on one of the nodes](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/50102/pull-kubernetes-e2e-kops-aws/98482/)

/test all
(let's figure out if there are possible flakes hidden and if this is isolated to AWS)

tnozicka commented:
pull-kubernetes-e2e-kops-aws falied to start node, remaining tests are green
/test all

smarterclayton commented on pkg/kubectl/cmd/get/get.go:
```diff
@@ -564,9 +572,17 @@ func (o *GetOptions) watch(f cmdutil.Factory, cmd *cobra.Command, args []string)
 	intr.Run(func() error {
-		_, err := watch.Until(0, w, func(e watch.Event) (bool, error) {
+		// FIXME: Switch to an informer (watchtools.UntilWithInformer) + dynamic client
+		// Pure WATCH will fail after some time but definitely on API timeout.
```
I'm not sure we want to use an informer, or that we want to remove watch-only as it is today.  We can certainly retry the watch from the last index until such a time as we can't.  The "endless watch" behavior of an informer is a different type of command (wait or observe).

tnozicka replied:
["So I would agree that in
certain circumstances (like tests, where the order of operations matters)
that a retry watcher that watches from the last RV received and errors if
it can't continue is a useful tool, BUT it should be really clear that most
people shouldn't use this.  That may imply we want to put it in its own
package, and clearly document when it is appropriate to use it."](https://github.com/kubernetes/kubernetes/pull/50102#issuecomment-321387829)

I believe that informer behavior should be the default option in CLI commands, unless you need to have the order of events guaranteed and fail. Informer also gives you the same guarantee for as long as the RetryWatcher would work.

Why `kubectl get` needs to fail if that guarantee is broken and not to try recovery?


> The "endless watch" behavior of an informer is a different type of command (wait or observe).

I don't think this applies in general (unless you meant it only for the case of `kubectl get`). Take the example of oc rollout status. If you implement it by RetryWatcher there will always be an option that it can fail, no way around that. Users re-implementing those in `oc observe` is not an option.

tnozicka replied:
or if the true intent of watch-only is --since=now? I guess we could filter events in the watch loop lower than an RV or something like that

tnozicka replied:
I am open to extracting those into and replacing it with a linked issue going with retrying Until for now.

liggitt replied:
> we could filter events in the watch loop lower than an RV

RV is opaque to clients... don't do client-side comparison

tnozicka replied:
To reach consensus sooner, I've dropped it from this PR and it uses only retry

tnozicka replied:
Actually it now uses only watchtools.UntilWithoutRetry because the unit tests are based on the unfortunate fact that we couldn't deal with watches being closed. Added a todo to replace it with retry (.Until) instead of an informer as asked.

smarterclayton commented on pkg/kubectl/cmd/rollout/rollout_status.go:
```diff
@@ -169,42 +183,67 @@ func (o *RolloutStatusOptions) Run() error {
+				return true, fmt.Errorf("watch failed: %v", e.Object)
+
+			default:
+				glog.Fatalf("unexpected event %#v", e)
```
You shouldn't be using log fatal and panic.  I'm not really sure what you're trying to do here.

tnozicka replied:
If my recollection is correct the thing is that glog.Fatal calls os.exit(255) eventually, the issue is compiler doesn't see it and thinks this branch has no return value. panic makes the compiler happy. I guess there could be a nicer way around it.

smarterclayton commented on pkg/kubectl/cmd/rollout/rollout_status.go:
```diff
@@ -169,42 +183,67 @@ func (o *RolloutStatusOptions) Run() error {
-			// Quit waiting if the rollout is done
-			if done {
-				return true, nil
+		_, err = watchtools.UntilWithInformer(ctx, lw, &unstructured.Unstructured{}, 0, preconditionFunc, func(e watch.Event) (bool, error) {
```
I'm not sure I like using `WithInformer` in the signature.  That's a detail of the implementation.  `UntilWithResync` is more appropriate.

I'm also uncomfortable with clients getting an `Added` event, because an informer is *not* a watch list.  We shouldn't be pretending that Added and Modified make sense with an informer loop, because they don't.

tnozicka replied:
> I'm not sure I like using WithInformer in the signature. That's a detail of the implementation. UntilWithResync is more appropriate.

UntilWithResync it is.

> I'm also uncomfortable with clients getting an Added event, because an informer is not a watch stream. We shouldn't be pretending that Added and Modified make sense with an informer loop, because they don't.

Could you elaborate on that idea?

> We shouldn't be pretending that Added and Modified make sense with an informer loop, because they don't.

I guess same could be said about Watch from RV "". You get Added event, although the object already existed. Why should this differ with an informer?

tnozicka replied:
@smarterclayton rename is done but I am gonna need you to explain your point with Added/Modified event in more detail though

smarterclayton commented on test/e2e/framework/deployment_util.go:
```diff
@@ -145,13 +148,16 @@ func WatchRecreateDeployment(c clientset.Interface, d *apps.Deployment) error {
-	if err != nil {
-		return err
+	watchFunc := func(sinceResourceVersion string) (watch.Interface, error) {
+		w, err := c.AppsV1().Deployments(d.Namespace).Watch(metav1.ListOptions{
```
Why did you stop using SingleObject here?

tnozicka commented:
/test all

tnozicka committed:
Update Bazel

tnozicka committed:
Update Godeps

tnozicka commented:
/milestone v1.12

tnozicka commented:
/kind bug
/priority important-soon

tnozicka commented:
Ready for reviews; all tests green.

This fixes significant number of bugs and performance issues.

/assign @wojtek-t 

PTAL @liggitt @deads2k @kubernetes/sig-api-machinery-pr-reviews @kubernetes/sig-apps-pr-reviews @kubernetes/sig-cli-pr-reviews

shyamjvs commented:
Related bugs:

- https://github.com/kubernetes/kubernetes/issues/56071 (iiuc this PR is a pre-requisite to fix that)
- https://github.com/kubernetes/kubernetes/issues/56064

Btw - @wojtek-t is on vacation until later part of August.

smarterclayton commented on test/e2e/apps/statefulset.go:
```diff
@@ -630,11 +643,17 @@ var _ = SIGDescribe("StatefulSet", func() {
+			ctx, cancel = context.WithTimeout(context.Background(), framework.StatefulSetTimeout)
+			defer cancel()
+			_, err = watchtools.Until(ctx, pl.ResourceVersion, watchFunc, func(event watch.Event) (bool, error) {
+				if event.Type == watch.Error {
```
The requirement that a caller now has to handle this is surprising and is going to break people who do the naive `Until` signature converse.n

I don't like that we changed the behavior of Until this dramatically.  I think this PR needs to introduce a new method with the desired behavior and keep the old behavior of Until.

tnozicka replied:
Actually that requirement has always been there as that event type was preexisting. 

But I admit we could handle it better. How about we catch status errors in UntilWithoutRetry by type and it returns them as Golang error, ending the loop?

smarterclayton commented on test/e2e/apps/statefulset.go:
```diff
@@ -734,10 +753,18 @@ var _ = SIGDescribe("StatefulSet", func() {
 			By("Waiting until stateful pod " + statefulPodName + " will be recreated and deleted at least once in namespace " + f.Namespace.Name)
-			w, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Watch(metav1.SingleObject(metav1.ObjectMeta{Name: statefulPodName}))
-			framework.ExpectNoError(err)
+			watchFunc := func(sinceResourceVersion string) (watch.Interface, error) {
```
This is a ton more boilerplate that callers are likely to get wrong.  While I don't know that I want to fix it here, I'm verye concerned about how likely someone is to get these methods wrong as a consequence of changing their code.

This seems like something that should be a generated method in the clients package that takes a `metav1.ListOptions` and does the right thing.  `podclientset.WatchSelector(opts metav1.ListOptions) func(rv string) (watch.Interface, error)`

tnozicka replied:
Yes, there is a bit more you need to write, I wouldn't call *a ton* since it's really just 2 lines - wrapping it in a lambda and setting RV option. There is a potential to get it wrong as always, but I've fixed it throughout the kube code base so given that people usually copy the pattern, they should be good and as always having a test for their code will make them aware if it works or not from the start.

Enhancing that experience with generated code could be nice follow up ;)

smarterclayton commented:
I do not like the size of this PR.  It needs to be broken out into a separate PR for the new machinery (new method, doesn't change existing signature) and the tests, then a set of smaller PRs that change individual areas of code.  I do not have confidence in this as a single monolithic PR, and I don't like the complexity that users have to deal with.

tnozicka commented on staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:
```diff
@@ -43,7 +43,7 @@ type StatusError struct {
 // for finer grained details.
 type APIStatus interface {
-	Status() metav1.Status
+	Status() *metav1.Status
```
Pointer doesn't always mean it's optional.

@liggitt had a nice idea to reuse apierrors.NewInternalError().Status() for passing it down in the event.Object except it needs to be a pointer to fulfill runtime.Object interface. Changing the APIStatus interface makes the Status() method on NewInternalError and others more usable.

tnozicka commented:
@smarterclayton 

Working on this in one PR proves that the apimachinery designs are valid and can cover all the use cases. (While fixing the kubectl I actually had several iterations on the machinery to cover it.) If splitting it into multiple PRs is what it takes, I can do that. **It will make merging it in time for 1.12 a lot harder though.**

I don't think the raised complexity for developers is terrible here. I think if I can get that watch.Error filtered out in UntilWithoutRetry you are left with just 2 additional lines wrapping it with lambda and setting RV. The fact that it actually makes it work after all those years should trump it without a blink of an eye, at least it does for me. We can have follow up to make the interface nicer for developers if it proves bug prone but we need to finally fix this for our users.

> doesn't change existing signature

I didn't touch the old Until when sending the PR but I got convinced it would be better to switch. Technically the signature is changing with the package move anyways, it's still available as UntilWithoutRetry which seems like a more descriptive name. 

Would it help to split the PR this way:
 - [ ] Move `"k8s.io/apimachinery/pkg/watch".Until` to `"k8s.io/client-go/tools/watch".UntilWithoutRetry`
 - [ ] Introduce `UntilWithSync`
 - [ ] Introduce `Until` (with retry)
 - [ ] Individual PRs to switch kubectl and other places to either `UntilWithSync` or `Until`

dims commented:
@tnozicka the issue https://github.com/kubernetes/kubernetes/issues/40224 is still marked v1.12

tnozicka commented:
> @tnozicka the issue #40224 is still marked v1.12

I intend to fix that one for 1.12 and scale - rest will go in 1.13 to limit the risks of initial bugs.

It was waiting for https://github.com/kubernetes/kubernetes/pull/66971 which just merged so I'll extract the rollout status fix into a PR today

soltysh commented on pkg/kubectl/cmd/rollout/rollout_status.go:
```diff
@@ -90,7 +103,7 @@ func NewCmdRolloutStatus(f cmdutil.Factory, streams genericclioptions.IOStreams)
 			cmdutil.CheckErr(o.Complete(f, args))
 			cmdutil.CheckErr(o.Validate(cmd, args))
-			cmdutil.CheckErr(o.Run())
+			cmdutil.CheckErr(o.Run(f))
```
That's unacceptable, we're reworking all the commands to get rid of this and you're adding it back. If you need dynamic client create it `Complete` and use in `Run`. Never do this!

tnozicka replied:
didn't know that, seemed like you've just removed unused argument in this case

I think you meant to comment on https://github.com/kubernetes/kubernetes/pull/67817, will fix it there

soltysh commented on pkg/kubectl/cmd/rollout/rollout_status.go:
```diff
@@ -155,11 +169,12 @@ func (o *RolloutStatusOptions) Run() error {
 	}
-	rv, err := meta.NewAccessor().ResourceVersion(obj)
+
+	dynamicClient, err := dynamic.NewForConfig(clientConfig)
```
This belongs in `Complete`.

soltysh commented on pkg/kubectl/cmd/run.go:
```diff
@@ -466,16 +471,41 @@ func (o *RunOptions) removeCreatedObjects(f cmdutil.Factory, createdObjects []*R
-	if err != nil {
-		return nil, err
+func waitForPod(podClient coreclient.PodsGetter, ns, name string, exitCondition watchtools.ConditionFunc) (*api.Pod, error) {
+	// TODO: expose the timeout
```
Don't leave a todo, expose it.

tnozicka replied:
out of scope for this change, if it wasn't till then - it's not the intend of these changes to fix the interface, just the machinery

soltysh commented on pkg/kubectl/cmd/wait/wait.go:
```diff
@@ -206,11 +208,11 @@ type ConditionFunc func(info *resource.Info, o *WaitOptions) (finalObject runtim
 
 // RunWait runs the waiting logic
 func (o *WaitOptions) RunWait() error {
+	// TODO: remove this unnecessary GET call, informers now take care about it all
```
If so - just do it.

soltysh reviewed (changes requested):
I left you some comments about kubectl code base changes

wojtek-t commented:
@tnozicka - now with 1.13 code cycle, can we resurrect this PR?

Given that parts of this were merged in the other two PRs you make, would you be able to rebase this one so that we can continue the work on it?
I think it's important to have this thing fixed finally :-)

wking commented:
You can check `kubectl rollout status` off in your initial comment now that #67817 [has landed][1].  And I really want to get `Until` back, so let me know if I can do anything to help get this resurrected :).

[1]: https://github.com/kubernetes/kubernetes/pull/67817#event-1825004872

wking commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,183 @@
+	if err != nil {
+		_ = rw.send(watch.Event{
+			Type:   watch.Error,
+			Object: apierrors.NewInternalError(fmt.Errorf("RetryWatcher: watcherFunc failed: %v", err)).Status(),
```
I dunno if this is an artifact of my local setup or not (I'm using k8s.io/apimachinery from kubernetes-1.12.2 and Go 1.10.3), but this is giving me:

```
...: cannot use ".../vendor/k8s.io/apimachinery/pkg/api/errors".NewInternalError(fmt.Errorf("RetryWatcher: watcherFunc failed: %v", err)).Status() (type ".../vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Status) as type runtime.Object in field value:
	".../vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Status does not implement runtime.Object (DeepCopyObject method has pointer receiver)
```

Instead, I had to use:

```go
status := apierrors.NewInternalError(fmt.Errorf("RetryWatcher: watcherFunc failed: %v", err)).Status()
_ = rw.send(watch.Event{
  Type:   watch.Error,
  Object: &status,
})
```

wking replied:
Also, can we have `fmt.Errorf("retry watcher:..."` to keep golint from complaining about:

    error strings should not be capitalized or end with punctuation or a newline

for low `-min_confidence`?  We just need something to give some context; we don't need to quote the Go type directly.

tnozicka replied:
old apimachinery; I can make the error start with lowercase

wking commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+// Done allows the caller to be notified when Retry watcher stops.
+func (rw *RetryWatcher) Done() <-chan struct{} {
```
Why do you need this?  Can't the caller notice this by closing `resultChan` (which you could do in a `defer` in `RetryWatcher.receive()` where you currently only close `doneChan`).  As it stands, the watcher closes `doneChan` but leaves `resultChan` alone, which leaves `Until` hung forever (unless someone else is watching for `doneChan` who can close `resultChan`).

wojtek-t commented:
@tnozicka - will you be able to rebase this PR - it would be great if we could make that happen in 1.14 timeframe

tnozicka commented:
@wojtek-t (missed your comment, didn't refresh) Sorry about the delay - paternity leave followed by KubeCon and end of the year shutdown piled up :(

Rebased RetryWatcher (Until) and and pushing it in https://github.com/kubernetes/kubernetes/pull/67350

Will follow by picking more commands to switch to UntilWithSync (as the experiment usage in `oc rollout status` seems to be working fine for 2 stable releases) and adding an experimental usage of the new `Until` backed by RetryWatcher somewhere.
